examples/cluster/plot_face_compress.py

```diff
@@ -1,76 +1,173 @@
+# 8 bits).
+#
+# Using 8-bit unsigned integer means that the image is encoded using 256 different
+# shades of gray, at most. We can check the distribution of these values.
+import matplotlib.pyplot as plt
+
+fig, ax = plt.subplots(ncols=2, figsize=(12, 4))
+
+ax[0].imshow(raccoon_face, cmap=plt.cm.gray)
+ax[0].axis("off")
+ax[0].set_title("Rendering of the image")
+ax[1].hist(raccoon_face.ravel(), bins=256)
+ax[1].set_xlabel("Pixel value")
+ax[1].set_ylabel("Number of pixels")
+ax[1].set_title("Distribution of the pixel values")
+_ = fig.suptitle("Original image of a raccoon face")
+
+# %%
+# Compression via vector quantization
+# -----------------------------------
+# The idea behind compression via vector quantization is to use a smaller number of
+# gray levels to represent the image. For instance, we can use 8 values instead of 256
+# values. Therefore, it means that we could efficiently use 1 bit instead of 8 bits to
+# encode a single pixel and therefore reduce the memory usage by a factor of 8. We will
+# later discuss about this memory usage.
+#
+# Encoding strategy
+# """""""""""""""""
+#
+# The compression can be performed by using a
+# :class:`~sklearn.preprocessing.KBinsDiscretizer`. We need to choose of a strategy
+# to define the 8 values. The simplest strategy is to define the 8 values equally
+# spaced that correspond to `strategy="uniform"`. From the previous histogram, we know
+# that this strategy is certainly not optimal.
+
+from sklearn.preprocessing import KBinsDiscretizer
+
+n_bins = 8
+encoder = KBinsDiscretizer(
+    n_bins=n_bins, encode="ordinal", strategy="uniform", random_state=0
+)
+compressed_raccoon_uniform = encoder.fit_transform(raccoon_face.reshape(-1, 1)).reshape(
+    raccoon_face.shape
+)
+
+fig, ax = plt.subplots(ncols=2, figsize=(12, 4))
+ax[0].imshow(compressed_raccoon_uniform, cmap=plt.cm.gray)
+ax[0].axis("off")
+ax[0].set_title("Rendering of the image")
+ax[1].hist(compressed_raccoon_uniform.ravel(), bins=256)
+ax[1].set_xlabel("Pixel value")
+ax[1].set_ylabel("Number of pixels")
+ax[1].set_title("Distribution of the pixel values")
+_ = fig.suptitle("Raccoon face compressed using 1-bit and a uniform strategy")
+
+# %%
+# Qualitatively, we can spot some small regions where we see the effect of the
+# compression (e.g. leaves on the bottom right corner). But after all, the resulting
+# image is still looking good.
+#
+# Looking a the distribution, we observe that all pixels values have been mapped to
+# 8 different values. But we can also see the amount of pixel in each bin is different.
+# This might not be the most effective strategy. We can check the correspondance between
+# the 8 values and the original pixel values.
+bin_edges = encoder.bin_edges_[0]
+bin_center = bin_edges[:-1] + (bin_edges[1:] - bin_edges[:-1]) / 2
+bin_center
+
+# %%
+plt.hist(raccoon_face.ravel(), bins=256)
+for center in bin_center:
+    plt.axvline(center, color="red")
```

Comment:
After checking, `twiny` is not a good idea since it dissociates the original `x` of the histogram and the one of the vertical line.

I just added some annotations instead.